module/programs/helper.js

```diff
@@ -976,6 +977,153 @@ module.exports = class ProgramsHelper {
+        
+        if ( !programData.length > 0 ) {
+          throw ({
+            status: httpStatusCode.bad_request.status,
+            message: constants.apiResponses.PROGRAM_NOT_FOUND
+          });
+        }
+        
+        let programUsersData = {};
+        // Fetch user profile information by calling sunbird's user read api.
+        // !Important check specific fields of userProfile.
+        let userProfile = await userService.profile(userToken, userId);
+        if (!userProfile.success || 
+            !userProfile.data ||
+            !userProfile.data.response ||
+            !userProfile.data.response.profileUserTypes ||
+            !userProfile.data.response.profileUserTypes.length > 0 ||
+            !userProfile.data.response.userLocations ||
+            !userProfile.data.response.userLocations.length > 0
+        ) {
+          throw ({
+            status: httpStatusCode.bad_request.status,
+            message: constants.apiResponses.PROGRAM_JOIN_FAILED
+          });      
+        } 
+        programUsersData = {
+          programId: programId,
+          userRoleInformation: data.userRoleInformation ? data.userRoleInformation : {},
+          userId: userId,
+          userProfile: userProfile.data.response
+        }
+        if( appName != "" ) {
+          programUsersData['appInformation.appName'] = appName;
+        }
+        if( appVersion != "" ) {
+          programUsersData['appInformation.appVersion'] = appVersion;
+        }
+
+        //For internal calls add consent using sunbird api
+        if(internalAccessToken !== ""){
+          let consent = {
+            "request": {
+              "consent": {
+                "status": constants.common.REVOKED,
+                "userId": userProfile.data.response.id,
+                "consumerId": userProfile.data.response.organisations.organisationId,
+                "objectId":  programId,
+                "objectType": constants.common.PROGRAM
+              }
+             }
+          }
+          let consentResponse = await userService.consent(userToken, consent)
+          if(!consentResponse.success){
+            throw {
+              message: constants.apiResponses.PROGRAM_JOIN_FAILED,
+              status: httpStatusCode.bad_request.status
+            }
+          }
+        }
+
+        //create or update query
+        const query = { 
+          programId: programId,
+          userId: userId
+        };
+
+        //Check data already present in db.
+        const programUsers = await programUsersHelper.find(
+          query,
+          ["_id"]
+        );
+        
+        // Data already present in programUsers collection, Require updation, else create new entry.
+        let joinProgram;
+        if( programUsers.length > 0 ) {
+          //Update the collection
+          let update = {};
+          update['$set'] = programUsersData;
+          if ( data.isResource ) {
+            update['$inc'] = { noOfResourcesStarted : 1 }
```

Comment:
@VISHNUDAS-tunerlabs we should update this count if after joining the program users start any solution